### ISSUE TITLE
HH\Asio\curl_exec does not work with hostnames if libcurl was built without ares

### ISSUE THREAD

**Issue by tchapi** [facebook/hhvm#5384]
Hi,

When I try to use `HH\Asio\curl_exec` in HHVM **3.7.0** in a very simple docker container, the result is always null (whatever the url), see this simple example, compared to a standard `curl_exec` that works as expected : 

```hack
<?hh

$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, "http://facebook.com");
$result = HH\Asio\curl_exec($ch)->getWaitHandle()->join();

var_dump($result); // String "" (0) — ??

$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, "http://facebook.com");
$result = curl_exec($ch);

var_dump($result); // bool(true) — seems legit
```

I don't know if the docker container has something to do with it, but I don't really see why it would, since curl behaves normally when called. I unfortunately don't have a server at hand to test in other conditions (i.e. _not in a docker container_).

My configuration seems fairly standard though : 

    root@eb4a1b4ebfc1:/# hhvm --version
    HipHop VM 3.7.0 (rel)
    Compiler: tags/HHVM-3.7.0-0-gc8baf9cd3cb603e030969bfe24634d5e85549915
    Repo schema: 1988ce75e6a571ce9c79c60fd7dc3ca8a5a5d403
    root@eb4a1b4ebfc1:/# curl --version
    curl 7.38.0 (x86_64-pc-linux-gnu) libcurl/7.38.0 OpenSSL/1.0.1k zlib/1.2.8 libidn/1.29 libssh2/1.4.3 librtmp/2.3

Any ideas ? Is there a known incompatibility with docker containers maybe ? Or something I am missing ?

Thanks a lot,

**Comment by jwatzman** [comment]
Does `curl http://facebook.com/` work inside your container? It returns an empty result on my Mac -- I think our load balancers reject requests from non-browser user-agents (as do many other sites, it seems, such as Wikipedia, Github, etc). Maybe try another URL?

**Comment by tchapi** [comment]
It works from the command line in the container. I tried as well with many different urls, but to no avail. I don't think it's a problem with `curl` since the regular `curl_exec()` works perfectly in exactly the same settings (see my test file above)

Best,

**Comment by jwatzman** [comment]
Ah yeah sorry missed that. @fredemmott or @sgolemon knows most about async curl.

**Comment by 5lava** [comment]
This has nothing to do with neither HHVM nor Docker. Request to http://facebook.com/ returns empty body (that's why it's ``String "" (0)``) with redirect to another URL. In my case I get 302 to https://facebook.com/ and then 301 to https://www.facebook.com/?_rdr.

You should add following to your code:
```php
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
```

**Comment by tchapi** [comment]
Once again the behaviour is the same on _any_ URL, may it redirect or not. As said in my previous comment, `curl_exec()` works as expected whereas `Asio\curl_exec()` does not.

The problem is the same if I add curl options such as `CURLOPT_FOLLOWLOCATION` or `CURLOPT_RETURNTRANSFER` for instance.

**Comment by 5lava** [comment]
What output do you get when you add ```curl_setopt($ch, CURLOPT_VERBOSE, 1)```?

**Comment by 5lava** [comment]
Also I'd recommend to request something simple like http://php.net/ to isolate the problem and minimize any possible SSL/redirection impact. I surprisingly caught segfault similar to #5355 while I was playing with your code.

**Comment by tchapi** [comment]
Hi,

`CURLOPT_VERBOSE` brings a little more info — apparently when called with **Asio**, `curl_exec` does not try to contact DNS servers : 

```hack
curl_setopt($ch, CURLOPT_URL, "http://example.com/");
$result = HH\Asio\curl_exec($ch)->getWaitHandle()->join();
```

gives : 

    hhvm_1  | * Hostname was NOT found in DNS cache  
    hhvm_1  | * Closing connection 0

But :

```hack
curl_setopt($ch, CURLOPT_URL, "http://example.com/");
$result = curl_exec($ch);
```

gives a normal result :

    hhvm_1  | * Hostname was NOT found in DNS cache
    hhvm_1  | *   Trying 93.184.216.34...
    hhvm_1  | * Connected to example.com (93.184.216.34) port 80 (#0)
    hhvm_1  | > GET / HTTP/1.1
    ( etc etc etc)
    hhvm_1  | * Connection #0 to host example.com left intact

So I'm confused. Why would the " Asio version " of `curl_exec` not contact the DNS servers ?
Thanks,

**Comment by 5lava** [comment]
I wouldn't say it's DNS related unless CURL reported something like ```Could not resolve: example.com (Domain name not found)```.

**Comment by tchapi** [comment]
Still the same in HHVM 3.7.1, any ideas @fredemmott or @sgolemon ?

**Comment by fredemmott** [comment]
Isolated repro (does not require docker) on 3.7.0 (not tested more recent versions):

```Hack
<?hh

async function byIP() {
  $ch = curl_init('http://93.184.216.34');
  curl_setopt($ch, CURLOPT_HTTPHEADER, ['Host: example.com']);
  $result = await HH\Asio\curl_exec($ch);
  return $result;
}

async function byHostname() {
  $ch = curl_init('http://example.com');
  $result = await HH\Asio\curl_exec($ch);
  return $result;
}

var_dump(HH\Asio\v(Vector { byIP(), byHostname() })->getWaitHandle()->join());
```

**Comment by fredemmott** [comment]
Narrowing it down a little:

```Hack
<?hh

// HH\Asio\curl_exec, but with new $async parameter
async function my_curl_exec(mixed $urlOrHandle, bool $async): Awaitable<string> {
  if (is_string($urlOrHandle)) {
    $ch = curl_init($urlOrHandle);
  } else if (is_resource($urlOrHandle) &&
             (get_resource_type($urlOrHandle) == "curl")) {
    $ch = $urlOrHandle;
  } else {
    throw new Exception(__FUNCTION__." expects string of cURL handle");
  }
  curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

  $mh = curl_multi_init();
  curl_multi_add_handle($mh, $ch);
  do {
    $active = 1;
    do {
      $status = curl_multi_exec($mh, $active);
    } while ($status == CURLM_CALL_MULTI_PERFORM);
    if (!$active) break;
    if ($async) {
      $select = await curl_multi_await($mh);
    } else {
      $select = curl_multi_select($mh);
    }
    if ($select == -1) break;
  } while ($status === CURLM_OK);
  $content = (string)curl_multi_getcontent($ch);
  curl_multi_remove_handle($mh, $ch);
  curl_multi_close($mh);
  return $content;
}

$ce1 = my_curl_exec('http://example.com', true);
$ce2 = my_curl_exec('http://example.com', false);
var_dump(HH\Asio\v(Vector{$ce1, $ce2})->getWaitHandle()->join());
```

**Comment by fredemmott** [comment]
Reproduces in a master OSS build. Does not repro in a master Facebook build.

**Comment by fredemmott** [comment]
curl_multi_await is returning -1 for the first round with DNS. Probably related: http://curl.haxx.se/libcurl/c/curl_multi_fdset.html

**Comment by fredemmott** [comment]
`If no file descriptors are set by libcurl, max_fd will contain -1 when this function returns. Otherwise it will contain the highest descriptor number libcurl set. When libcurl returns -1 in max_fd, it is because libcurl currently does something that isn't possible for your application to monitor with a socket and unfortunately you can then not know exactly when the current action is completed using select(). You then need to wait a while before you proceed and call curl_multi_perform anyway. How long to wait? We suggest 100 milliseconds at least, but you may want to test it out in your own particular conditions to find a suitable value.`

The synchronous version doesn't work in PHP5. It only works in HHVM because we switched to using curl_multi_wait() instead of curl_multi_fdset() for the synchronous version.

**Comment by fredemmott** [comment]
This depend on the machine's build of curl: it works on our test machines because our build of CuRL use asyn-ares.c, but Ubuntu's default one uses asyn-thread.c (both files from libcurl upstream), and asyn-thread.c has a no-op implementation of Curl_resolver_getsock()

**Comment by fredemmott** [comment]
Fix:

```
diff --git a/hphp/runtime/ext/curl/ext_curl.php b/hphp/runtime/ext/curl/ext_curl.php
index f4cbd7c..559b8c7 100644
--- a/hphp/runtime/ext/curl/ext_curl.php
+++ b/hphp/runtime/ext/curl/ext_curl.php
@@ -359,6 +359,7 @@ async function curl_exec(mixed $urlOrHandle): Awaitable<string> {

   $mh = curl_multi_init();
   curl_multi_add_handle($mh, $ch);
+  $sleep_ms = 10;
   do {
     $active = 1;
     do {
@@ -366,7 +367,18 @@ async function curl_exec(mixed $urlOrHandle): Awaitable<string>
     } while ($status == CURLM_CALL_MULTI_PERFORM);
     if (!$active) break;
     $select = await curl_multi_await($mh);
-    if ($select == -1) break;
+    /* If cURL is built without ares support, DNS queries don't have a socket
+     * to wait on, so curl_multi_await() (and curl_select() in PHP5) will return
+     * -1, and polling is required.
+     */
+    if ($select == -1) {
+      await SleepWaitHandle::create($sleep_ms * 1000);
+      if ($sleep_ms < 1000) {
+        $sleep_ms *= 2;
+      }
+    } else {
+      $sleep_ms = 10;
+    }
   } while ($status === CURLM_OK);
   $content = (string)curl_multi_getcontent($ch);
   curl_multi_remove_handle($mh, $ch);
```

As a work-around, you can copy the implementation of HH\Asio\curl_exec from hphp/runtime/ext/curl/ext_curl.php and create a new function in your own code that has the same change.

**Comment by fredemmott** [comment]
Fix up internally: D2274514

**Comment by tchapi** [comment]
@fredemmott Thanks a lot !
